File: elfin_teleop/scripts/elfin_keys_to_pose.py
#!/usr/bin/env python
import time
import copy
import roslib
import rospy
import tf.transformations
import numpy as np
from std_msgs.msg import String
from sensor_msgs.msg import JointState
from geometry_msgs.msg import Pose, Point, Quaternion

# ---------------- CLASS ----------------
class ElfinKeyToPoseNode():
    def __init__(self):
        print('ElfinKeyToPoseNode: initializing node')

        self.ready = False
        self.joint_states = None

        # ROS Infrastructure
        topic = 'keys'
        self.sub_keys = rospy.Subscriber(topic, String, self.keys_callback)
        topic = 'pose_cmd'
        self.pub_pose_cmd = rospy.Publisher(topic, Pose, queue_size=10)
        topic = 'joint_states'
        self.sub_joint_states = rospy.Subscriber(topic, JointState, self.joint_states_callback)

        # Wait to get ready
        print('Waiting for joint_states...')
        while (not self.ready):
            pass
        print('Received first joint_states message')
        
        self.desired_position, self.desired_orientation = self.fkine(self.joint_states.position)
        self.desired_pose = Pose(self.desired_position, self.desired_orientation)

        rospy.sleep(1)


    def keys_callback(self, msg):
        if not self.ready:
            return

        i = 0.1
        key = msg.data

        # Position
        if(key=="q"):
            self.desired_position.x += i
        elif(key=="a"):
            self.desired_position.x -= i
        
        elif(key=="w"):
            self.desired_position.y += i
        elif(key=="s"):
            self.desired_position.y -= i
     
        elif(key=="e"):
            self.desired_position.z += i
        elif(key=="d"):
            self.desired_position.z -= i

        # Rotation wrt base
        elif(key=="r"):
            rot = tf.transformations.quaternion_about_axis(i, (1, 0, 0))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply(rot, [o.x, o.y, o.z, o.w])
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])
        elif(key=="f"):
            rot = tf.transformations.quaternion_about_axis(-i, (1, 0, 0))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply(rot, [o.x, o.y, o.z, o.w])
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])
     
        elif(key=="t"):
            rot = tf.transformations.quaternion_about_axis(i, (0, 1, 0))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply(rot, [o.x, o.y, o.z, o.w])
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])
        elif(key=="g"):
            rot = tf.transformations.quaternion_about_axis(-i, (0, 1, 0))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply(rot, [o.x, o.y, o.z, o.w])
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])
     
        elif(key=="y"):
            rot = tf.transformations.quaternion_about_axis(i, (0, 0, 1))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply(rot, [o.x, o.y, o.z, o.w])
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])
        elif(key=="h"):
            rot = tf.transformations.quaternion_about_axis(-i, (0, 0, 1))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply(rot, [o.x, o.y, o.z, o.w])
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])

        # Rotation wrt ee
        elif(key=="u"):
            rot = tf.transformations.quaternion_about_axis(i, (1, 0, 0))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply([o.x, o.y, o.z, o.w], rot)
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])
        elif(key=="j"):
            rot = tf.transformations.quaternion_about_axis(-i, (1, 0, 0))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply([o.x, o.y, o.z, o.w], rot)
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])
     
        elif(key=="i"):
            rot = tf.transformations.quaternion_about_axis(i, (0, 1, 0))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply([o.x, o.y, o.z, o.w], rot)
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])
        elif(key=="k"):
            rot = tf.transformations.quaternion_about_axis(-i, (0, 1, 0))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply([o.x, o.y, o.z, o.w], rot)
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])
     
        elif(key=="o"):
            rot = tf.transformations.quaternion_about_axis(i, (0, 0, 1))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply([o.x, o.y, o.z, o.w], rot)
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])
        elif(key=="l"):
            rot = tf.transformations.quaternion_about_axis(-i, (0, 0, 1))
            o = self.desired_orientation
            result = tf.transformations.quaternion_multiply([o.x, o.y, o.z, o.w], rot)
            self.desired_orientation = Quaternion(result[0], result[1], result[2], result[3])

        elif(key=="z"):
            self.desired_position, self.desired_orientation = self.fkine([0, 0, 0, 0, 0, 0])
        elif(key=="x"):
            self.desired_position, self.desired_orientation = self.fkine(self.joint_states.position)
        elif(key=="m"):
            self.desired_position, self.desired_orientation = self.fkine([1.57, 0.78, -1.57, 0, 0.78, 0.0])

        self.desired_pose.position = self.desired_position
        self.desired_pose.orientation = self.desired_orientation

        self.pub_pose_cmd.publish(self.desired_pose)


    def joint_states_callback(self, msg):
        self.joint_states = msg
        self.ready = True


    def dh(self, th, d, alpha, a):
        mat = np.array([[np.cos(th), -np.cos(alpha)*np.sin(th),  np.sin(alpha)*np.sin(th), a*np.cos(th)],
                        [np.sin(th),  np.cos(alpha)*np.cos(th), -np.sin(alpha)*np.cos(th), a*np.sin(th)],
                        [         0,             np.sin(alpha),             np.cos(alpha),            d],
                        [         0,                         0,                         0,            1]])
        return mat


    def fkine(self, q):
        T1 = self.dh(q[0]          , 0.22,  np.pi/2, 0   )
        T2 = self.dh(q[1] + np.pi/2, 0   ,  np.pi  , 0.38)
        T3 = self.dh(q[2] + np.pi/2, 0   ,  np.pi/2, 0   )
        T4 = self.dh(q[3] + np.pi  , 0.42,  np.pi/2, 0   )
        T5 = self.dh(q[4]          , 0   , -np.pi/2, 0   )
        T6 = self.dh(q[5] + np.pi  , 0.18,  0      , 0   )
        pose = T1.dot(T2).dot(T3).dot(T4).dot(T5).dot(T6)

        trans = tf.transformations.translation_from_matrix(pose)
        orient = tf.transformations.quaternion_from_matrix(pose)
        return Point(trans[0], trans[1], trans[2]), Quaternion(orient[0], orient[1], orient[2], orient[3])


if __name__ == '__main__':
    print('starting elfin_keys_to_pose')
    rospy.init_node('elfin_keys_to_pose', disable_signals=True)
    node = ElfinKeyToPoseNode()

    try:
        rospy.spin()
    except:  # rospy.ROSInterruptException:
        print('caught exception')
    print('exiting')

File: elfin_teleop/src/elfin_keys_to_pose.cpp
#include "elfin_teleop/elfin_keys_to_pose.h"

ElfinKeysToPoseNode::ElfinKeysToPoseNode(){
	// Initialize Variables
	ready = false;

	// ROS Infrastructure
	joint_state_sub = nh.subscribe(JOINT_STATE_TOPIC, 1, &ElfinKeysToPoseNode::jointStateCallback, this);
	pose_cmd_pub = nh.advertise<geometry_msgs::Pose>(POSE_CMD_TOPIC, 5);
	keys_sub = nh.subscribe(KEYS_TOPIC, 1, &ElfinKeysToPoseNode::keysCallback, this);

	// Wait to get ready
	ROS_INFO("Waiting for joint_states...");
	while(!ready)
		ros::spinOnce();
	ROS_INFO("Received first joint_states message");
	
	// Initialize desire position, orientation and pose	
	forwardKinematics(current_joint_state.position, desired_position, desired_orientation);
	desired_pose.position = desired_position;
	desired_pose.orientation = desired_orientation;
}

ElfinKeysToPoseNode::~ElfinKeysToPoseNode(){
	nh.shutdown();
}

void ElfinKeysToPoseNode::keysCallback(const std_msgs::String::ConstPtr& msg){
	if (!ready)
		return;

	double i = 0.05;
	tf2::Vector3 axis;
	tf2::Quaternion rot, current_quat, result;
	std::vector<double> zero_vector, main_vector = {1.57, 0.78, -1.57, 0, 0.78, 0.0};
	zero_vector.resize(6);

	std::string key = msg->data.c_str();
	char key_char = (char) key[0];
	switch (key_char) {
	// Position
	case 'q':
		desired_position.x += i;
		break;
	case 'a':
		desired_position.x -= i;
		break;
	case 'w':
		desired_position.y += i;
		break;
	case 's':
		desired_position.y -= i;
		break;
	case 'e':
		desired_position.z += i;
		break;
	case 'd':
		desired_position.z -= i;
		break;
	// Rotation wrt base
	case 'r':
		axis = tf2::Vector3(1, 0, 0);
		rot = tf2::Quaternion(axis, i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = rot * current_quat;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	case 'f':
		axis = tf2::Vector3(1, 0, 0);
		rot = tf2::Quaternion(axis, -i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = rot * current_quat;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	case 't':
		axis = tf2::Vector3(0, 1, 0);
		rot = tf2::Quaternion(axis, i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = rot * current_quat;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	case 'g':
		axis = tf2::Vector3(0, 1, 0);
		rot = tf2::Quaternion(axis, -i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = rot * current_quat;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	case 'y':
		axis = tf2::Vector3(0, 0, 1);
		rot = tf2::Quaternion(axis, i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = rot * current_quat;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	case 'h':
		axis = tf2::Vector3(0, 0, 1);
		rot = tf2::Quaternion(axis, -i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = rot * current_quat;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	// Rotation wrt End Efector
	case 'u':
		axis = tf2::Vector3(1, 0, 0);
		rot = tf2::Quaternion(axis, i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = current_quat * rot;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	case 'j':
		axis = tf2::Vector3(1, 0, 0);
		rot = tf2::Quaternion(axis, -i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = current_quat * rot;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	case 'i':
		axis = tf2::Vector3(0, 1, 0);
		rot = tf2::Quaternion(axis, i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = current_quat * rot;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	case 'k':
		axis = tf2::Vector3(0, 1, 0);
		rot = tf2::Quaternion(axis, -i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = current_quat * rot;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	case 'o':
		axis = tf2::Vector3(0, 0, 1);
		rot = tf2::Quaternion(axis, i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = current_quat * rot;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	case 'l':
		axis = tf2::Vector3(0, 0, 1);
		rot = tf2::Quaternion(axis, -i);
		current_quat = tf2::Quaternion(desired_orientation.x, desired_orientation.y, desired_orientation.z, desired_orientation.w);
		result = current_quat * rot;
		result.normalize();
		desired_orientation.x = result.x();
		desired_orientation.y = result.y();
		desired_orientation.z = result.z();
		desired_orientation.w = result.w();
		break;
	// Main and zero positions
	case 'z':
		forwardKinematics(zero_vector, desired_position, desired_orientation);
		break;
	case 'x':
		forwardKinematics(current_joint_state.position, desired_position, desired_orientation);
		break;
	case 'm':
		forwardKinematics(main_vector, desired_position, desired_orientation);
		break;
	}

	desired_pose.position = desired_position;
	desired_pose.orientation = desired_orientation;

	pose_cmd_pub.publish(desired_pose);
}

void ElfinKeysToPoseNode::jointStateCallback(const sensor_msgs::JointState& msg){
	current_joint_state = msg;
	ready = true;
}

void ElfinKeysToPoseNode::denavitHartenberg(double th, double d, double alpha, double a, Eigen::Matrix4d &dh){
	dh<<cos(th), -cos(alpha)*sin(th),  sin(alpha)*sin(th), a*cos(th),
    	sin(th),  cos(alpha)*cos(th), -sin(alpha)*cos(th), a*sin(th),
    	      0,          sin(alpha),          cos(alpha),         d,
    	      0,                   0,                   0,         1;
}

void ElfinKeysToPoseNode::forwardKinematics(std::vector<double> &q, geometry_msgs::Point &trans, geometry_msgs::Quaternion &orient){
	// Initialize Variables for calculation
	Eigen::Matrix4d T1, T2, T3, T4, T5, T6, fkine;
	Eigen::Matrix3d orient_eigen;
	Eigen::Vector3d trans_eigen;
	Eigen::Quaterniond quat;

	// Obtain Pose of End Efector
	denavitHartenberg(q[0]         , 0.22,  M_PI_2, 0   , T1);
	denavitHartenberg(q[1] + M_PI_2, 0   ,  M_PI  , 0.38, T2);
	denavitHartenberg(q[2] + M_PI_2, 0   ,  M_PI_2, 0   , T3);
	denavitHartenberg(q[3] + M_PI  , 0.42,  M_PI_2, 0   , T4);
	denavitHartenberg(q[4]         , 0   , -M_PI_2, 0   , T5);
	denavitHartenberg(q[5] + M_PI  , 0.18,  0     , 0   , T6);
	fkine = T1 * T2 * T3 * T4 * T5 * T6;

	// Extract position and orientation
	trans_eigen<< fkine(0,3), fkine(1,3), fkine(2,3);
	orient_eigen<<	fkine(0,0), fkine(0,1), fkine(0,2),
					fkine(1,0), fkine(1,1), fkine(1,2), 
					fkine(2,0), fkine(2,1), fkine(2,2);
	quat = Eigen::Quaterniond(orient_eigen);

	// Return msgs
	trans.x = trans_eigen(0);
	trans.y = trans_eigen(1);
	trans.z = trans_eigen(2);
	orient.x = quat.x();
	orient.y = quat.y();
	orient.z = quat.z();
	orient.w = quat.w();
}

int main(int argc, char** argv)
{
	ros::init(argc, argv, "elfin_keys_to_pose");

	ElfinKeysToPoseNode node;
	ros::spin();

	return 0;
}